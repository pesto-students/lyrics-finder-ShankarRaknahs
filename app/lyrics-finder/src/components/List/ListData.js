import { MdPerson } from 'react-icons/md';

const ListData = ({ category, data }) => {
  return (
    <section>
      <div className='category'>
        <button className='header'>{category}</button>
        <hr className='ruler' />
      </div>
      <ul>
        {data.map((item) => (
          <li key={item.id} className='list'>
            <div className='column'>
              <div className='title'>{item.title_short}</div>
              <MdPerson className='icon' />
              <span className='artist'>{item.artist.name}</span>
            </div>
            <span className='album'>{item.album.title}</span>
            <hr className='divider' />
          </li>
        ))}
      </ul>
    </section>
  );
};

export default ListData;
