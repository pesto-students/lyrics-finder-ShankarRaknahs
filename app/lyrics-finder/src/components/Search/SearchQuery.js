import { useState, useCallback } from 'react';
import ListData from '../List/ListData';
import _ from 'lodash';
import getData from '../../api/getData';
import CONFIG from '../../config/app.config';
import Filters from '../Filters/Filters';

const SearchQuery = () => {
  const [searchObj, setSearchObj] = useState({
    query: '',
    data: [],
    isLoading: false,
  });

  const [apiStatus, setApiStatus] = useState({
    code: CONFIG.HTTP_SUCCESS_CODE,
    statusText: '',
  });

  const onSearch = ({ target: { value } }) => {
    setSearchObj({ isLoading: true }, debouncedCallback(value));
    setApiStatus({ code: CONFIG.HTTP_SUCCESS_CODE, statusText: '' });
  };

  const fetchData = (query) => {
    const url = new URL(query, CONFIG.SEARCH_API);
    query.trim() !== ''
      ? getData(url)
          .then((suggestions) => {
            setSearchObj({
              query: query,
              data: suggestions['data'],
              isLoading: false,
            });
          })
          .catch((error) => {
            setSearchObj({
              data: [],
              isLoading: false,
            });
            setApiStatus({ code: error.status, statusText: error.statusText });
          })
      : setSearchObj({ query: query, data: [], isLoading: false });
  };

  const debounced = _.debounce(fetchData, CONFIG.SEARCH_DEBOUNCE_TIME);
  const debouncedCallback = useCallback(debounced, []);

  return (
    <section className='search'>
      <div className='query'>
        <input
          type='text'
          className='search-bar'
          placeholder='Search for Albums or Songs'
          onChange={onSearch}
        />
      </div>
      {apiStatus.code === CONFIG.HTTP_SUCCESS_CODE ? (
        <div className='result'>
          {searchObj.isLoading ? (
            <p className='loader'> Loading ...</p>
          ) : searchObj.data.length > 0 ? (
            <ListData category='#Your Songs' data={searchObj.data} />
          ) : searchObj.query !== '' ? (
            <div> {CONFIG.NO_RESULT_FOUND} </div>
          ) : (
            <Filters filterCategories={CONFIG.FILTER_CATEGORIES} />
          )}
        </div>
      ) : (
        <pre className='error-response'>{` ${CONFIG.ERROR_MESSAGES.apiError} 
        ${apiStatus.code} - ${apiStatus.statusText} 
        ${CONFIG.ERROR_MESSAGES.action}`}</pre>
      )}
    </section>
  );
};

export default SearchQuery;
