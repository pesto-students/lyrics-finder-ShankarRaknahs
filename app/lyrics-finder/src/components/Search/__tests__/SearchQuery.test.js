import React from 'react';

import renderer from 'react-test-renderer';

// import { render } from '@testing-library/react';
// import userEvent from '@testing-library/user-event';
import SearchQuery from '../SearchQuery';
import ListData from '../../List/ListData';

import Enzyme, { shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJSON from 'enzyme-to-json';
Enzyme.configure({ adapter: new Adapter() });
const { mount } = Enzyme;

describe('<SearchQuery />', () => {
  it('render correctly (snapshot)', () => {
    const tree = renderer.create(<SearchQuery />);
    expect(toJSON(tree)).toMatchSnapshot();
  });

  it('render correctly using Enzyme (snapshot)', () => {
    const tree = mount(<SearchQuery />);
    expect(toJSON(tree)).toMatchSnapshot();
  });
});

// describe('Children <ListData/>', () => {
//   it('is present', () => {
//     const [searchObj, setSearchObj] = React.useState({
//       query: '',
//       data: [{ test }],
//       isLoading: false,
//     });

//     const viewport = shallow(<SearchQuery />);
//     expect(
//       viewport.find(
//         <ListData category='#Search Results' data={searchObj.data} />
//       ).length
//     ).toBe(1);
//   });
// });
