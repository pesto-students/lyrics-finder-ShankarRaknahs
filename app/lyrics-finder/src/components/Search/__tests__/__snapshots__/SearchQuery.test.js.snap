// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchQuery /> render correctly (snapshot) 1`] = `undefined`;

exports[`<SearchQuery /> render correctly using Enzyme (snapshot) 1`] = `
<SearchQuery>
  <section
    className="search"
  >
    <div
      className="query"
    >
      <input
        className="search-bar"
        onChange={[Function]}
        placeholder="Search for Albums or Songs"
        type="text"
      />
    </div>
    <div
      className="result"
    />
  </section>
</SearchQuery>
`;
